src/rebuild/scan.c

```diff
@@ -636,26 +658,96 @@ placement_check(uuid_t co_uuid, vos_iter_entry_t *ent, void *data)
 	return rc;
 }
 
-struct rebuild_iter_arg {
-	ds_iter_cb_t	callback;
-	void		*arg;
-};
+static int
+rebuild_container_scan_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+			  vos_iter_type_t type, vos_iter_param_t *iter_param,
+			  void *data, unsigned *acts)
+{
+	struct rebuild_scan_xarg	*xarg = data;
+	struct rebuild_scan_arg		*arg = xarg->arg;
+	struct rebuild_tgt_pool_tracker *rpt = arg->rpt;
+	vos_iter_param_t		param = { 0 };
+	struct vos_iter_anchors		anchor = { 0 };
+	daos_handle_t			coh;
+	int				rc;
+
+	/* resync DTXs' status firstly. */
+	if (uuid_compare(xarg->co_uuid, entry->ie_couuid) == 0) {
+		D_DEBUG(DB_REBUILD, DF_UUID" already scan\n",
+			DP_UUID(xarg->co_uuid));
+		return 0;
+	}
+
+	rc = dtx_resync(iter_param->ip_hdl, rpt->rt_pool_uuid, entry->ie_couuid,
+			rpt->rt_rebuild_ver, true);
+	if (rc) {
+		D_ERROR(DF_UUID" dtx resync failed: rc %d\n",
+			DP_UUID(rpt->rt_pool_uuid), rc);
+		return rc;
+	}
+
+	rc = vos_cont_open(iter_param->ip_hdl, entry->ie_couuid, &coh);
+	if (rc != 0) {
+		D_ERROR("Open container "DF_UUID" failed: rc = %d\n",
+			DP_UUID(entry->ie_couuid), rc);
+		return rc;
+	}
+
+	memset(&param, 0, sizeof(param));
+	param.ip_hdl = coh;
+	param.ip_epr.epr_lo = 0;
+	param.ip_epr.epr_hi = DAOS_EPOCH_MAX;
+	param.ip_flags = VOS_IT_FOR_REBUILD;
+	uuid_copy(xarg->co_uuid, entry->ie_couuid);
+	rc = vos_iterate(&param, VOS_ITER_OBJ, false, &anchor,
+			 rebuild_obj_scan_cb, xarg);
+	vos_cont_close(coh);
+
+	/* Since dtx_resync might yield, let's reprobe anyway */
```

Comment:
The comment is some confused, the rebuild_obj_scan_cb() itself also may be yield.

Reply:
yes, obj_scan_cb can be yield as well, it will set YIELD flag.